File: Programas/Aguilar García Miguel Ángel -Tarea 8. part 1/InterpolaciondeLagrange.m
%Miguel Angel Aguilar Garcia
%Metodos numericos
%Fecha de entrega: Martes 26 de Agosto de 2014
%Interpolacion de Lagrange
q=1;
while q==1
clear all;
clc;

disp('Interpolacion de Lagrange');
      fprintf('Para ingresar la tabla de valores x/y hacerlo por fichero como si fuera una\nmatriz\ndel lado izquierdo ''x'' y del lado derecho ''y''\n');
      fprintf('Desea:\n1. Hacer un fichero nuevo o abrir para modificar\n2. Abrir un fichero existente');
      fprintf('\n\n******NOTA IMPORTANTE:Cuando se acaben de poner todos los elementos de la matriz\nse debera anadir un enter\n(es decir,situar el cursor debajo del ultimo renglon)\nde lo contrario saldran errores,\nademas se debera poner un espacio finalizado un renglon,\nde lo contrario saldran errores');         
      q=input('\nElige una opcion: ');
      
switch q
               case 1
      fprintf('\n1. Hacer un fichero nuevo o abrir para modificar');  
      fprintf('\n\nNota: Para seguir con el proceso se debe guardar\nlas modificaciones que se hicieron en el blog de notas\ny cerrarlo');     
      d=input('\n\nIntroduce el nombre del fichero que desea crear o modificar (nombre.txt): ','s');
      fclose('all');
      f1=fopen(d,'r+');
      fclose(f1);
      system(d);
      f1=fopen(d,'r');
      [A,cont]=fscanf(f1,'%f');
      fclose(f1);
      
      cont1=1;
      cont2=1;
for i=1:cont
if(mod(i,2)==0)
yv(cont1)=A(i);
cont1=cont1+1;
end
if(mod(i,2)~=0)
xv(cont2)=A(i);
cont2=cont2+1;
end
end

for i=1:cont1-1
    fprintf('%.2f\t%.2f\n',xv(i),yv(i));  
end

      q=0;
               case 2
      fprintf('\n2. Abrir un fichero existente');       
      d=input('\nIntroduce el nombre del fichero que desea abir (nombre.txt): ','s');
      fclose('all');
      system(d);
      f1=fopen(d,'r');
      [A,cont]=fscanf(f1,'%f');
      fclose(f1);
      
      cont1=1;
      cont2=1;
      
for i=1:cont
if(mod(i,2)==0)
yv(cont1)=A(i);
cont1=cont1+1;
end
if(mod(i,2)~=0)
xv(cont2)=A(i);
cont2=cont2+1;
end
end

for i=1:cont1-1
    fprintf('%.2f\t%.2f\n',xv(i),yv(i));  
end
      q=0;
      
      
    otherwise
               a=0;
               fprintf('\n\nERROR.Opcion no valida\n\n');
               pause;
               q=1;
end

end

syms x;
n=cont1-1;
prod=1;
sum=0;
for i=1:n
    prod=1;
    for j=1:n
    
        if(i~=j)
        prod=prod*(x-xv(j))/(xv(i)-xv(j));
        end
    end
prod=prod*yv(i);
sum=sum+prod;
end

disp(' ');
disp('La funcion es: ');
disp(sum);

ezplot(sum,[1.5,2]);
hold on;
grid ON;

disp('1. Repetir programa');
    disp('o presiona cualquier otro numero para terminar');
    des=input('Selecciona la opción: ');
    
    if des==1
        InterpolaciondeLagrange;
    end

File: Programas/Aguilar García Miguel Ángel -Tarea 8. part 1/Interpolacion de Lagrange.cpp
/*Miguel Angel Aguilar Garcia*/
/*Metodos numericos*/
/*Fecha de entrega: Martes 26 de Agosto de 2014*/
/*Interpolacion de Lagrange*/

#include<stdio.h>
#include<stdlib.h>
#include<conio.h>
#include<math.h>

double prod=1,sum=0,xval,xv[1000],yv[1000];

int i=1,j=1,n=0,m=0,r=0,e=0,t=0,q=0,x=0,nn=0,mm=0;
char op,a,d[40],cc; 
double ma[1000][1000];
FILE *f1;

int main()
{
      
do
{
      do
      {
      q=0;
      system("cls");
      fclose(f1);
      printf("Interpolacion de Lagrange\n\n");
      printf("Para ingresar la tabla de valores x/y hacerlo por fichero como si fuera una\nmatriz\ndel lado izquierdo 'x' y del lado derecho 'y'\n");
      printf("Desea:\
               \n1. Hacer un fichero nuevo o abrir para modificar\
               \n2. Abrir un fichero existente");
      printf("\n\n******NOTA IMPORTANTE:Cuando se acaben de poner todos los elementos de la matriz\nse debera anadir un enter\n(es decir,situar el cursor debajo del ultimo renglon)\nde lo contrario saldran errores,\nademas se debera poner un espacio finalizado un renglon,\nde lo contrario saldran errores");         
      printf("\n\nElige una opcion: ");
      scanf("%d",&q);
      fflush(stdin);
      switch(q)
      {
               case 1:
                    {
      a=0;
      system("cls");  
      printf("1. Hacer un fichero nuevo o abrir para modificar");  
      printf("\n\nNota: Para seguir con el proceso se debe guardar\nlas modificaciones que se hicieron en el blog de notas\ny cerrarlo\n\n");       
      printf("\nIntroduce el nombre del fichero que desea crear o modificar (nombre.txt): ");
      scanf("%s",&d);
      fflush(stdin);
      
      f1=fopen(d,"a");
      fclose(f1);
      system(d);/*Abrir blog de notas*/
      f1=fopen(d,"r");
      printf("\n\nContinuar si ya ha guardado el blog de notas\n\n");
      system("pause");
      q=0;
                        break;
                    }
               case 2:
                    {

      a=0;
      system("cls");  
      printf("2. Abrir un fichero existente");       
      printf("\nIntroduce el nombre del fichero que desea abir (nombre.txt): ");
      scanf("%s",&d);
      fflush(stdin);
      f1=fopen(d,"r");
      q=0;
                        break;
                    }
               default:
               {
               a=0;
               printf("\n\nERROR.Opcion no valida\n\n");
               system("pause");
               q=1;
               break;
               }
      }
      
      }while(q==1);

      if(f1!=NULL)
      {
       a=0;/*Se ponen los valores para evitar errores al momento de la repeticion*/
       i=1;
       j=1;
       n=0;
       m=0;
       r=0;
       e=0;
       t=0;    
       system("cls");  
       printf("Se ha abierto correctamente");
         
  while((cc=fgetc(f1))!=EOF)
 {                                    
   if(cc=='\n')
   {
   e++;/*Contador de filas*/      
   }                                                  
 }
 
fseek(f1,0,SEEK_SET);
  
   while((cc=fgetc(f1))!='\n')
 {                                    
   if((cc=='\t')||(cc==' '))
   {
   t++;/*Contador de columnas*/          
   }                                                 
 }
fseek(f1,0,SEEK_SET);

  while((cc=fgetc(f1))!=EOF)
 {                                      
   if((cc=='\t')||(cc==' '))
   {
   r++;/*Contador de elementos totales*/       
   }                                                     
 }
printf("\nElementos totales= %d",r);
printf("\nFilas= %d",e);
printf("\nColumnas= %d",t);

m=e+1;
n=t+1;
fseek(f1,0,SEEK_SET);
x=0;
       for(i=1;i<m;i++)
       {
                           for(j=1;j<n;j++)
                           {
                                            
                                            if(a!=EOF)
                                            {
                                            fscanf(f1,"%lf",&ma[i][j]);/*Asignar elementos a matriz*/
                                            fflush(stdin);
                                            x++;
                                            }
                           } 
       }

fclose(f1);

printf("\n\nElementos totales= %d",x); 
nn=n-1;
mm=m-1;
              if(x!=r)
              {
              printf("\n\nERROR. Posiblemente las filas no estan completas o le faltan elementos,\
              \nverifique que en todas las filas el numero de columnas es igual\
              \nsi faltan elementos complete con ceros\n\n");
              system("pause");
              exit(0);
              }
                     
       printf("\nSu matriz es:");  
       printf("\n");  
       for(i=1;i<m;i++)
       {
                           for(j=1;j<n;j++)
                           {
                           printf("%lf ",ma[i][j]);/*Imprimir matriz*/     
                           }
                   printf("\n");        
       }
       
printf("\n\nVerifique que su matriz es la correcta,\nsi hay algun error verifique su matriz,\nsi persiste el error, introducir desde cero su matriz");
system(d);/*Abrir blog de notas*/ 
      }
     
      else
      {
       printf("\nError al abrir el archivo\n");
       system("pause");
      }
      
/*******************************************************************************************************************/      
/*Hacer los vectores 'x' y 'y'*/
for(i=1;i<=mm;i++)
{
    xv[i]=ma[i][1];
    yv[i]=ma[i][2];
    
}
/******************************************************************************/
system("cls");

printf("Interpolacion de Lagrange\n\n");
               
printf("Inserte un valor de 'x' a evaluar: ");
scanf("%lf",&xval);
fflush(stdin);
/******************************************************************************/
n=mm;
prod=1;
sum=0;
for(i=1;i<=n;i++)
{
    prod=1;
    for(j=1;j<=n;j++)
    {
        if(i!=j)
        {
        prod*=(xval-xv[j])/(xv[i]-xv[j]);
        }
    }
prod*=yv[i];
sum+=prod;
}
/******************************************************************************/
printf("\nEl valor de 'y' para ese valor de 'x' es: %lf",sum);
/******************************************************************************/ 

printf("\n\nDesea repetir (S/N): ");
scanf("%c",&op);
fflush(stdin);
}
while(op=='s'||op=='S');

printf("\n\n");
system("pause");
}
